### ISSUE TITLE
FluxInpaintPipeline overrides pixels outside the mask

### ISSUE THREAD

**Issue by Clement-Lelievre** [huggingface/diffusers#9867]
### Describe the bug

When inpainting (with diffusers==0.31.0 and torch==2.4.1) using `FluxInpaintPipeline`, I get some pixels outside the mask (and pretty far away from the mask border) that are overrided whereas the mask at their indices was black.
With either flux schnell or dev.

Is this expected?

I could paste the input image on the output image using the mask to get rid of this, but I'd like to check first if this is a bug or not.

Here is my input image and mask:

![vampire_input_rgb](https://github.com/user-attachments/assets/502e0a66-0351-4563-b693-7b804644efa1)
![vampire_mask](https://github.com/user-attachments/assets/158d5359-e229-4e3c-a317-a143b19c97a1)

Here is the kind of diff I get:

![diff](https://github.com/user-attachments/assets/11dca503-5edd-49cb-8a90-6d147897aaf3)

### Reproduction

I am using a small, square (224*224) image (see above), and a non-rectangular mask that has values only in {0, 255}.

```python
from diffusers import FluxInpaintPipeline

pipe = FluxInpaintPipeline.from_pretrained("black-forest-labs/FLUX.1-schnell", torch_dtype=torch.bfloat16, device='cuda')
prompt = "skull"
image= # above, convert it to RGB
mask = # above,  convert it to RGB
image = pipe(prompt=prompt, image=source, mask_image=mask, height=224, width=224, num_inference_steps=2, guidance_scale=3.5, strength=0.76).images[0]  # the latter params are probably not necessary to reproduce the issue
image.save("flux_inpainting.png")

# now visualize the pixel diff between the output you get and the input image, you should see non-zero pixels outside the mask, particularly on the right part
```

### System Info

- 🤗 Diffusers version: 0.31.0.dev0 or 0.31.0
- Platform: Linux-5.15.0-1048-aws-x86_64-with-glibc2.31
- Running on Google Colab?: No
- Python version: 3.12.0
- PyTorch version (GPU?): 2.4.1+cu121 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Huggingface_hub version: 0.25.0
- Transformers version: 4.44.2
- Accelerate version: 0.34.2
- PEFT version: 0.12.0
- Bitsandbytes version: not installed
- Safetensors version: 0.4.5
- xFormers version: not installed
- Accelerator: NVIDIA A10G, 23028 MiB

### Who can help?

@asomoza @sayakpaul

**Comment by sayakpaul** [comment]
PyTorch 3.2.0? That sounds like invalid.

**Comment by Clement-Lelievre** [comment]
> PyTorch 3.2.0? That sounds like invalid.

torch==2.4.1

**Comment by yiyixuxu** [comment]
cc @asomoza

**Comment by SahilCarterr** [comment]
That can be done but it would require changing Pipelines including all ControlNets and Inpainting pipelines because they all have the same difference.
> I believe it's due to VAE encode/decode. It's not a bijective method. If confirmed, a blend with the decoded image at the [end](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_inpaint.py#L1008) will solve the difference.

**Comment by crapthings** [comment]
> I believe it's due to VAE encode/decode. It's not a bijective method. If confirmed, a blend with the decoded image at the [end](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_inpaint.py?rgh-link-date=2024-11-06T16%3A22%3A44Z#L1008) will solve the difference.

can this be apply to controlnet inpainting?
is this method same as https://huggingface.co/docs/diffusers/api/image_processor#diffusers.image_processor.VaeImageProcessor.apply_overlay

**Comment by christopher5106** [comment]
Right, you can use apply_overlay on the final PIL image in both inpaint and controlnet inpaint to correct the small VAE differences. But I would apply the mask directly on the image before postprocess just after [decode](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_controlnet_inpainting.py#L1184)

**Comment by christopher5106** [comment]
```python
            ...
            image = self.vae.decode(latents, return_dict=False)[0]
            image = (1 - init_mask) * init_image + init_mask * image
            image = self.image_processor.postprocess(image, output_type=output_type)
            ...
```

**Comment by crapthings** [comment]
> ```python
>             ...
>             image = self.vae.decode(latents, return_dict=False)[0]
>             image = (1 - init_mask) * init_image + init_mask * image
>             image = self.image_processor.postprocess(image, output_type=output_type)
>             ...
> ```

What are the advantages of doing it this way? Compared to directly applying overlay? Also, can this be applied to control networks?

i'm trying to apply this into controlnet by copy this line

```
image = self.vae.decode(latents, return_dict=False)[0]
```

to pag controlnet pipeline, but i got a grayish result

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd.py#L1317

**Comment by putdanil** [comment]
I have issues too, parts of the original image (black parts) show up on the inpainted image.

Removing this line "solves" this, but seems to damage the diffusion process
`latents = (1 - init_mask) * init_latents_proper + init_mask * latents`


![fluxtest011124 (19)](https://github.com/user-attachments/assets/fc0bd19d-b1d8-4f02-a942-eb453111ddcf)

**Comment by sayakpaul** [comment]
Cc: @yiyixuxu @asomoza

**Comment by asomoza** [comment]
there seems to be a misunderstanding of issues here, so I'll try to clear them here.

The inpainting pipeline uses a binary mask to replace part of the image, this is not subjective and it uses math (tensors) to do it, so there's no "it changes other parts of the image" issues like people are suggesting here, there are other factors that are making this appear to happen.

The vae encode/decode is not s lossless process, each time you encode the image to feed it to the model and then decode it to convert it back, you're degrading the original image or the final one if you don't do anything else. This is worse if you do this process several times, the second or third time the image gets considerable worse.

This is an inpainting pipeline only and not a model, there's no Flux model training for the inpainting task so don't expect it to work  like with an inpainting model or controlnet, the model doesn't know how to blend the original image with the generation, so it's just luck if it generates something that fits or not with the rest of the image. For flux you can se the "refill" model and pipeline better.

The best method to make this work is with using an inpainting model and contrary to the suggestion proposed before, to blend the new generated part with the **original** image before the vae encode process, so this means not to do it in the latent space but with the pillow images, with this you prevent the first loss you get with the encoding. 

This is not something we do automatically inside the pipeline because this depends on the use case of the final user, so it's left to the user or app to decide what to do with the inpainting results. For example, sometimes is good to do a histogram matching to prevent different saturations and colors between them, maybe sharpen them or blend them with a special technique like a pixel per pixel blending (soft inpainting) on the borders or something similar.

**Comment by crapthings** [comment]
@asomoza 

what is different between controlnet inpainting vs inpainting model?

**Comment by asomoza** [comment]
sorry for the late answer @crapthings, I missed this somehow.

> what is different between controlnet inpainting vs inpainting model?

That question is not simple or fast to answer, but the short answer is the training.

A longer but not that complete answer would be:

The inpainting model is just adding more inputs to the model and finetune it with the correct images and text so the whole model learns how to do the inpainting, this mostly means that you restrict this to just one model with the caveat that you can get the difference in the weights and try to apply them to other models that share the same base model, this is what is done for SD 1.5 and SDXL models.

Controlnet is injecting additional neural layers to the model and train only those in the special conditions you want, that's why you can use for example depth maps or canny images and it works, this also have the advantage that you can set the scale and when this conditioning starts and ends, so controlnet is a lot more versatile than a simple fine tune, it also allows the same controlnet model to be used with any other model if it has the same architecture. You can refer to the [controlnet paper](https://arxiv.org/pdf/2302.05543) if you want to learn more.

Controlnet inpainting is just a controlnet trained with images, text and masks, same as the inpainting finetune.

Both methods achieve the same thing at the end but controlnet seems to be easier to train, more versatile and better at the job.

**Comment by asomoza** [comment]
since there's no more follow ups to this, I'll close it, if someone else post another issue or question I'll reopen it.
